pkg/kubelet/kubelet.go

```diff
@@ -1734,6 +1726,143 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+	// not get invoked again
+	if runningPod != nil {
+		// we kill the pod with the specified grace period since this is a termination
+		if gracePeriod != nil {
+			klog.V(4).InfoS("Pod terminating with grace period", "pod", klog.KObj(pod), "podUID", pod.UID, "gracePeriod", *gracePeriod)
+		} else {
+			klog.V(4).InfoS("Pod terminating with grace period", "pod", klog.KObj(pod), "podUID", pod.UID, "gracePeriod", nil)
+		}
+		if err := kl.killPod(pod, *runningPod, gracePeriod); err != nil {
+			kl.recorder.Eventf(pod, v1.EventTypeWarning, events.FailedToKillPod, "error killing pod: %v", err)
+			// there was an error killing the pod, so we return that error directly
+			utilruntime.HandleError(err)
+			return err
+		}
+		klog.V(4).InfoS("Pod termination stopped all running orphan containers", "pod", klog.KObj(pod), "podUID", pod.UID)
+		return nil
+	}
+
+	// In order to safely terminate a pod, we need to know the set of sandboxes that the last
+	// syncPod method created as well as running containers - but we also wish to do so with
+	// the lowest latency possible, which means we don't want the pod worker to have to wait
+	// to call syncTerminatingPod until container status is recalculated. Instead, the first
+	// invocation of syncTerminatingPod on a given pod may send a nil podStatus to allow us
+	// to more quickly guarantee shutdown of everything the last syncPod created.
+	if podStatus == nil {
+		var err error
+		podStatus, err = kl.pleg.WaitForUpdatedStatus(&kubecontainer.Pod{ID: pod.UID, Namespace: pod.Namespace, Name: pod.Name})
+		if err != nil {
+			klog.ErrorS(err, "Unable to refresh pod status prior to killing pod", "pod", klog.KObj(pod), "podUID", pod.UID)
+			return err
+		}
+	}
+
+	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus)
+	if podStatusFn != nil {
+		podStatusFn(&apiPodStatus)
+	}
+	kl.statusManager.SetPodStatus(pod, apiPodStatus)
+
+	if gracePeriod != nil {
+		klog.V(4).InfoS("Pod terminating with grace period", "pod", klog.KObj(pod), "podUID", pod.UID, "gracePeriod", *gracePeriod)
+	} else {
+		klog.V(4).InfoS("Pod terminating with grace period", "pod", klog.KObj(pod), "podUID", pod.UID, "gracePeriod", nil)
+	}
+	p := kubecontainer.ConvertPodStatusToRunningPod(kl.getRuntime().Type(), podStatus)
+	if err := kl.killPod(pod, p, gracePeriod); err != nil {
+		kl.recorder.Eventf(pod, v1.EventTypeWarning, events.FailedToKillPod, "error killing pod: %v", err)
+		// there was an error killing the pod, so we return that error directly
+		utilruntime.HandleError(err)
+		return err
+	}
+
+	// Read the most recent container status from the PLEG so that we maximize the chance that we'll
+	// see the status of terminated containers (since we could race with other cleanup processes on the
+	// node in low resource situations).
+	// TODO: have KillPod return the terminal status of stopped containers
+	podStatus, err := kl.pleg.WaitForUpdatedStatus(&kubecontainer.Pod{ID: pod.UID, Namespace: pod.Namespace, Name: pod.Name})
```

Comment:
Offline discussed.

Only this call to the container runtime to get pod status is required for correctness.

The other direct container runtime pod status call and pod cache update are mostly for optimization: reducing latency. We probably want to solve the race condition and ensure correctness first.

The latency reduction can be done in a separate change, and we can evaluate different options, e.g.:
1) Make `GetPodStatus` lightweight, so that we are more open to the breakglass call for latency sensitive scenario to bypass 1s PLEG relist;
2) Make PLEG event driven, so that status can be updated (no control flow critical) faster.

Those may need more thoughts, and can be a separate change. :)

Reply:
Pushed review comment 1 with updates to match our discussion.